### ISSUE TITLE
Use enums instead of constants

### ISSUE THREAD

**Issue by disconnect3d** [trailofbits/manticore#1343]
There are many places where we use pure strings instead of enums and later we do comparisons on those strings which are rather slow.

This is probably not the most bottleneck here and there (I haven't profiled it), but it is still worth to change it for better.

Ethereum:
* Gas setting: 
  * Tx type/sort {CREATE, CALL, DELEGATECALL}
  * Tx result type {STOP, RETURN, SELFDESTRUCT, THROW, TXERROR, REVERT}

Native: 
* Arch (i386, amd64, armv7) and arm64 in the future

TODO steps:
* [ ] Find out all cases like this and report it here to be confirmed
* [ ] Fix each of the case individually - with its own PR, also please attach a small benchmark so we can see it helped for the better

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 100.0 DAI (100.0 USD @ $1.0/DAI)  attached to it as part of the Ethereum Foundation fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/trailofbits/manticore/1343/2121).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/trailofbits/manticore/1343/2121).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $38,686.65 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 1 month, 1 week ago.
Please review their action plans below:


**1) [alogishetty](https://gitcoin.co/alogishetty) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=approve&worker=alogishetty) | [reject worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=reject&worker=alogishetty))_.**

Will go through the code to identify where string values are used for comparison and implement enums for better performance.
**2) [jaeseokan94](https://gitcoin.co/jaeseokan94) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=approve&worker=jaeseokan94) | [reject worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=reject&worker=jaeseokan94))_.**

1. Run grep -r "in {\"[A-Z]\{2,\}" .     &lt;- meaning, recursively find string patterns that starts with in followed by { and two or more capital letters. Why this? Because, we want to change constant comparison to enum. 

Result: 
./manticore/platforms/evm.py:        if sort not in {"CREATE", "CALL", "DELEGATECALL"}:
./manticore/platforms/evm.py:        if self.result in {"RETURN", "STOP", "SELFDESTRUCT"}:
./manticore/platforms/evm.py:            assert self.result in {"TXERROR", "REVERT", "THROW"}
./manticore/platforms/evm.py:        if result in {"RETURN", "REVERT"}:
./manticore/platforms/evm.py:        elif result in {"STOP", "THROW", "SELFDESTRUCT"}:
./manticore/platforms/evm.py:        if self.result in {"STOP", "RETURN", "SELFDESTRUCT"}:
./manticore/platforms/evm.py:            assert self.result in {"THROW", "TXERROR", "REVERT"}
./manticore/platforms/evm.py:        if sort not in {"CALL", "CREATE", "DELEGATECALL", "CALLCODE"}:
./manticore/utils/emulate.py:            if reg in {"FS", "GS"}:
./manticore/utils/emulate.py:        if reg in {"FS"}:
./manticore/ethereum/manticore.py:        if tx.result in {"SELFDESTRUCT", "REVERT", "THROW", "TXERROR"}:
./manticore/ethereum/manticore.py:        elif tx.result in {"RETURN", "STOP"}:

I can also find similar cases to find out if there any constant set to be changed to enum. 

2. Change one constant at a time. Let's say, I change, {"RETURN", "STOP"}, to enum, then use "grep" keyword to find all the constants being used, and change them to enum. May add a line in files to import enum. 

One commit should be self-contained and only has one constant set changes. 

3. Repeat, get reviews commits by commits. 


Question.
How come changing string comparison to enum make performance better? I don't get it. 
I understand that changing it to enum is a lot better since spelling mistakes will be caught by IDE or exception easily, so we still need to make this changes, but I don't understand the impacts of enum in performance.
**3) [seenivasanseeni](https://gitcoin.co/seenivasanseeni) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=approve&worker=seenivasanseeni) | [reject worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=reject&worker=seenivasanseeni))_.**

1. list all the cases of constant using.
2. get feedback
3. iterate 1 and 2 untill approved
4. for each use case, fix it.
**4) [arun0930](https://gitcoin.co/arun0930) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=approve&worker=arun0930) | [reject worker](https://gitcoin.co/issue/trailofbits/manticore/1343/2121?mutate_worker_action=reject&worker=arun0930))_.**

0x7F4B0fc7c8BD0fe9e472B9F66e89EEC1BE52089f

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/trailofbits/manticore/1343/2121).

**Comment by nanspro** [comment]
@disconnect3d  is this issue still available?
I have worked before on a similar issue where i implemented enums in typescript so i think i 'll be able to finish it quick.
I have good python experience also understanding codebase won't be a problem!!

**Comment by janus** [comment]
Is this issue still open? I am interested

**Comment by ehennenfent** [comment]
Yes! If you can figure out how to apply via GitCoin, please do so.

**Comment by janus** [comment]
@ehennenfent  I gave applied .... approve please

**Comment by ehennenfent** [comment]
@janus I've sent an email to the gitcoin people asking them to approve work on this issue. I'm afraid it's in their hands now.

**Comment by Man-Jain** [comment]
@ehennenfent Can you please tell who have you approved to work on this issue.

**Comment by ehennenfent** [comment]
@Man-Jain At the time I reached out, only @janus had applied to work on this issue. As far as I know, we're still waiting on @vs77bb or perhaps @ceresstation to approve things via GitCoin. Trail of Bits doesn't have any direct control over approval since we're not the ones who funded the issue.

**Comment by Man-Jain** [comment]
Ohh! Thanks for the info.

**Comment by ehennenfent** [comment]
@mkosowsk Any chance we can get this approved? I know you're not the original funder of this issue but I see that we've worked with you on previous issues.

**Comment by mkosowsk** [comment]
@ceresstation mind taking a look?

**Comment by seeni-dev** [comment]
can you take a look at my application?

**Comment by zjwjing** [comment]
I would like to work on this issue. I have analyzed the codebase and found the following string constants that should be replaced with enums:

## Ethereum (manticore/ethereum/manticore.py)

### Transaction Types (sort)
- Line 730: `"CREATE"`
- Line 800: `"CALL"`
- Line 950: `"CREATE", "CALL"`
- Line 953: `sort == "CREATE"`
- Line 997: `sort == "CREATE"`
- Line 1475: `tx.sort == "CREATE"`
- Line 1501: `current_transaction.sort == "CREATE"`
- Line 1561: `current_transaction.sort == "CREATE"`

### Transaction Results (result)
- Line 1476: `tx.result == "RETURN"`
- Line 1758: `last_tx.result in {"REVERT", "THROW", "TXERROR"}`

## Native (manticore/native/cpu/arm.py)
- Line 601: `machine = "armv7"`

## Plan
1. Create enum classes for TxType and TxResult in a new file `manticore/ethereum/enums.py`
2. Replace string comparisons with enum comparisons
3. Add benchmark to show performance improvement

I will submit a PR with the fix and benchmark results.
